### PR TITLE
NDEV-1767 retention 

### PR DESCRIPTION
**Pull request by valiksinev**

### CODE REVIEW

**Comment by valiksinev** [comment]
> 1. Provide test config file
> 2. Use logger and console output at the same time. You can use fastlog.
> 3. Introduce config path to the config.

config.yaml  is attached
fast_log added
log_file_name added to config

File: v2-advanced/clickhouse-retention/src/client.rs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+            }
+        }
+
+        if !tasks.is_empty() {
+            for task in tasks {
+                task.await.unwrap()?;
+            }
+        }
+        let execution_time = Instant::now().duration_since(time_start);
+        println!("execution time: {} sec", execution_time.as_secs());
```
[the remark this answers is not in the document]
Comment (valiksinev) [reply v2-advanced/clickhouse-retention/src/client.rs]: decimal part  - it unnecessary information for this task

File: v2-advanced/clickhouse-retention/src/client.rs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        let row_count = self.row_count(slot_start, slot_end, &partition).await?;
+        if row_count == 0 {
+            return Err(clickhouse::error::Error::Custom("no rows found to copy".to_string()));
+        }
+        println!("{} rows should be copied", row_count);
+
+        println!("slots to copy: {:?}..{:?}", slot_start, slot_end);
+
+        let mut slot = slot_start;
+        let range = 24000;  // 216000:9 = 24000
```
[the remark this answers is not in the document]
Comment (valiksinev) [reply v2-advanced/clickhouse-retention/src/client.rs:146]: added to config

File: v2-advanced/clickhouse-retention/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    let logger: &'static Logger = fast_log::init(Config::new().console().file_split(
+        &config.log_file_name,
+        LogSize::KB(512),
+        RollingType::All,
+        LogPacker {},
+    ))
+        .expect("Failed to initialize fast_log");
+    logger.set_level(LevelFilter::Info);
+
+    info!("Start copiyng to older_account");
```
Comment (Deniskore) [inline v2-advanced/clickhouse-retention/src/main.rs:29]: Fix typo please.

File: v2-advanced/clickhouse-retention/src/client.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    let elapsed = Instant::now().duration_since(start).as_secs_f64();
+
+    info!("slots {}..{} done, task took {:.2} sec" , first_slot,last_slot, elapsed);
+    Ok(())
+}
+
+impl <'a>Connection<'a> {
+    pub fn new(config: &'a Config) -> Self {
+        let id = rand::thread_rng().gen_range(0..config.clickhouse_url.len());
+        let url = config.clickhouse_url.get(id).unwrap();
```
Comment (Deniskore) [inline v2-advanced/clickhouse-retention/src/client.rs:68]: We can replace it with expect()

File: v2-advanced/clickhouse-retention/src/config.rs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    let clickhouse_user = env::var("NEON_DB_CLICKHOUSE_USER")
+        .map(Some)
+        .unwrap_or(None);
+
+    let clickhouse_password = env::var("NEON_DB_CLICKHOUSE_PASSWORD")
+        .map(Some)
+        .unwrap_or(None);
+
+    let copy_partition_offset: u64 =  env::var("COPY_PARTITION_OFFSET")
+        .expect("COPY_PARTITION_OFFSET should be set").parse::<u64>().unwrap();
```
Comment (Deniskore) [inline v2-advanced/clickhouse-retention/src/config.rs:69]: Please replace unwrap calls with expect.

**Review by Deniskore — CHANGES_REQUESTED** [review]

File: v2-advanced/clickhouse-retention/src/client.rs (hunk 4 above)
Comment (valiksinev) [reply v2-advanced/clickhouse-retention/src/client.rs:68]: fixed

File: v2-advanced/clickhouse-retention/src/config.rs (hunk 5 above)
Comment (valiksinev) [reply v2-advanced/clickhouse-retention/src/config.rs:69]: fixed

**Review by Deniskore — APPROVED** [review]

**Pull request merged**